File: src/main/java/net/adventuregame/mobs/Pathfinding.java
package net.adventuregame.mobs;

import com.chidozie.core.terrains.Terrain;
import org.joml.Vector3f;
import java.util.*;

public class Pathfinding {

    public static List<Vector3f> findPath(Vector3f start, Vector3f target, Terrain terrain) {
        PriorityQueue<Node> openSet = new PriorityQueue<>(Comparator.comparingDouble(n -> n.fCost));
        Map<Vector3f, Node> allNodes = new HashMap<>();

        Node startNode = new Node(start, null, 0, start.distance(target));
        openSet.add(startNode);
        allNodes.put(start, startNode);

        while (!openSet.isEmpty()) {
            Node current = openSet.poll();

            if (current.position.equals(target)) {
                return constructPath(current);
            }

            for (Vector3f neighbor : getNeighbors(current.position, terrain)) {
                float gCost = current.gCost + current.position.distance(neighbor);
                float hCost = neighbor.distance(target);
                Node neighborNode = allNodes.getOrDefault(neighbor, new Node(neighbor, null, Float.MAX_VALUE, hCost));

                if (gCost < neighborNode.gCost) {
                    neighborNode.gCost = gCost;
                    neighborNode.fCost = gCost + hCost;
                    neighborNode.parent = current;
                    openSet.add(neighborNode);
                    allNodes.put(neighbor, neighborNode);
                }
            }
        }

        return Collections.emptyList(); // No path found
    }

    private static List<Vector3f> getNeighbors(Vector3f position, Terrain terrain) {
        List<Vector3f> neighbors = new ArrayList<>();
        float stepSize = 5.0f; // Grid resolution
        float terrainHeight = terrain.getHeightOfTerrain(position.x, position.z);

        for (float dx = -stepSize; dx <= stepSize; dx += stepSize) {
            for (float dz = -stepSize; dz <= stepSize; dz += stepSize) {
                if (dx == 0 && dz == 0) continue;
                Vector3f newPos = new Vector3f(position.x + dx, terrainHeight, position.z + dz);
                neighbors.add(newPos);
            }
        }
        return neighbors;
    }

    private static List<Vector3f> constructPath(Node node) {
        List<Vector3f> path = new ArrayList<>();
        while (node != null) {
            path.add(node.position);
            node = node.parent;
        }
        Collections.reverse(path);
        return path;
    }

    private static class Node {
        Vector3f position;
        Node parent;
        float gCost, fCost;

        Node(Vector3f position, Node parent, float gCost, float hCost) {
            this.position = position;
            this.parent = parent;
            this.gCost = gCost;
            this.fCost = gCost + hCost;
        }
    }
}

File: src/main/kotlin/net/adventuregame/mobs/Pathfinding.kt
package net.adventuregame.mobs

import com.adv.core.terrains.Terrain
import org.joml.Vector3f
import java.util.*
import java.util.concurrent.Executors
import java.util.function.ToDoubleFunction

object Pathfinding {
    private val executor = Executors.newFixedThreadPool(4){ runnable ->
        Thread(runnable).apply { isDaemon = true }
    }

    data class Vec3Key(val x: Float, val y: Float, val z: Float) {
        constructor(v: Vector3f) : this(v.x, v.y, v.z)
    }

    fun findPath(start: Vector3f, target: Vector3f?, terrain: Terrain): MutableList<Vector3f>? {
        try {
       //     println("Finding Path")
            val openSet =
                PriorityQueue<Node>(Comparator.comparingDouble(ToDoubleFunction { n: Node? -> n!!.fCost.toDouble() }))
            val allNodes: MutableMap<Vec3Key, Node> = HashMap<Vec3Key, Node>()

            if (target == null) {
         //       println("NO TARGET")
                return mutableListOf()
            }

            val startNode = Node(position = start, parent = null, gCost = 0f, hCost = start.distance(target))
           // println("Created Node with UUID=${startNode.id}")
            openSet.add(startNode)
            allNodes.put(Vec3Key(start), startNode)

            while (!openSet.isEmpty()) {
                val current = openSet.poll()

                if (current.position.distance(target) < 5f) {
             //       println("Constructing Path")
                    return constructPath(current)
                }

                for (neighbor in getNeighbors(current.position, terrain)) {
                    val gCost = current.gCost + current.position.distance(neighbor)
                    val hCost = neighbor.distance(target)
                    val neighborNode = allNodes.getOrDefault(
                        Vec3Key(neighbor),
                        Node(position = neighbor, parent = null, gCost = Float.MAX_VALUE, hCost = hCost)
                    )

                    if (gCost < neighborNode.gCost) {
                        neighborNode.gCost = gCost
                        neighborNode.fCost = gCost + hCost
                        neighborNode.parent = current
                        if (!openSet.contains(neighborNode)) {
               //             println("Adding Node ${neighborNode.id} to openSet at ${neighborNode.position}")
                            openSet.add(neighborNode)
                        }
                        allNodes.put(Vec3Key(neighbor), neighborNode)
                    }
                }
            }
            //println("No conditions met, returning empty list")
            return mutableListOf() // No path found
        } catch (e: Exception) {
            //println("Exception Caught in FindPath!")
            return mutableListOf()
        }
    }

    private fun getNeighbors(position: Vector3f, terrain: Terrain): MutableList<Vector3f> {
        val neighbors: MutableList<Vector3f> = ArrayList<Vector3f>()
        val stepSize = 5.0f // Grid resolution

        var dx = -stepSize
        while (dx <= stepSize) {
            var dz = -stepSize
            while (dz <= stepSize) {
                if (dx == 0f && dz == 0f) {
                    dz += stepSize
                    continue
                }
                val newX = position.x + dx
                val newZ = position.z + dz
                val newY = terrain.getHeightOfTerrain(newX, newZ)
                if (newY.isNaN()) println("WARNING: terrain height NaN at $newX, $newZ")
                val newPos = Vector3f(newX, newY, newZ)
                neighbors.add(newPos)
                dz += stepSize
            }
            dx += stepSize
        }
        return neighbors
    }

    private fun constructPath(node: Node?): MutableList<Vector3f>? {
        //println("Constructing PATH!!")
        var node = node
        val path: MutableList<Vector3f>? = ArrayList()
        while (node != null) {
            path!!.add(node.position)
            node = node.parent
        }
        //println("Reversing Path")
        path!!.reverse()
        return path
    }

    private class Node(
        val id: UUID = UUID.randomUUID(), // Add this
        var position: Vector3f,
        var parent: Node?,
        var gCost: Float,
        hCost: Float
    ) {
        var fCost: Float = gCost + hCost

        override fun toString(): String {
            return "Node(id=$id, pos=$position, g=$gCost, f=$fCost)"
        }
    }

    fun findPathAsync(
        start: Vector3f,
        target: Vector3f?,
        terrain: Terrain,
        onComplete: (List<Vector3f>) -> Unit
    ) {
        println("Find Path Async Called")
        println("Executor isShutdown=${executor.isShutdown}, isTerminated=${executor.isTerminated}")
        executor.execute {
            println("Executing Executor")
            try {
                val result = findPath(start, target, terrain) ?: emptyList()
                println(">>> [Kotlin] Pathfinding result: ${result.size}")
                onComplete(result)
            } catch (e: Exception) {
                println(">>> [Kotlin] Pathfinding CRASHED: ${e.message}")
                e.printStackTrace()
            }
        }
    }
}